**Submitted by: Group G4**->(Umesha B K,Raja Ram Chaudhary,Gowthami R)<br>
**2020-06-24**

![](airindia.png)
<h1 style="text-align:center;"> Air Fare prediction Analysis</h1>

 Import necessaries library

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

Read given data by help of pandas function 

In [2]:
data = pd.read_csv('C:\\Users\\Raja\\Desktop\\project\\project2020\\air_fair.csv')

## Exploratory Data Analysis (EDA)<a id="3"></a>

Show top 5 records

In [3]:
data.head()

InvoiceDate  NetFare    ProductType ItineraryType
0     1/12/18 11:43  2160.00  Other Product      Domestic
1     1/12/18 11:43     0.49  Other Product      Domestic
2     1/12/18 13:01  6550.00            Air      Domestic
3     1/12/18 13:10  6550.00            Air      Domestic
4  30/11/2018 14:30  5497.00            Air      Domestic

InvoiceDate converted int datetime format

In [4]:
data['InvoiceDate']=pd.to_datetime(data['InvoiceDate'])

Extraction of Date, Year, Month, Day, Hour & Minute from InvoiceDate columns

In [5]:
data['Date']=data['InvoiceDate'].dt.date
data['Year']=data['InvoiceDate'].dt.year
data['Month']=data['InvoiceDate'].dt.month
data['Day']=data['InvoiceDate'].dt.day
data['Hour']=data['InvoiceDate'].dt.hour
data['Minute']=data['InvoiceDate'].dt.minute
data

InvoiceDate  NetFare    ProductType ItineraryType        Date  \
0      2018-01-12 11:43:00  2160.00  Other Product      Domestic  2018-01-12   
1      2018-01-12 11:43:00     0.49  Other Product      Domestic  2018-01-12   
2      2018-01-12 13:01:00  6550.00            Air      Domestic  2018-01-12   
3      2018-01-12 13:10:00  6550.00            Air      Domestic  2018-01-12   
4      2018-11-30 14:30:00  5497.00            Air      Domestic  2018-11-30   
...                    ...      ...            ...           ...         ...   
278461 2019-10-06 20:45:00     0.00            Air      Domestic  2019-10-06   
278462 2019-10-06 21:39:00  9073.00            Air      Domestic  2019-10-06   
278463 2019-10-06 22:18:00  2367.00            Air      Domestic  2019-10-06   
278464 2019-10-06 22:30:00  6572.00            Air      Domestic  2019-10-06   
278465 2019-10-06 22:30:00  5118.00            Air      Domestic  2019-10-06   

          Year  Month   Day  Hour  Minute  
0       2018.0    1.0  12.0  11.0    43.0  
1       2018.0    1.0  12.0  11.0    43.0  
2       2018.0    1.0  12.0  13.0     1.0  
3       2018.0    1.0  12.0  13.0    10.0  
4       2018.0   11.0  30.0  14.0    30.0  
...        ...    ...   ...   ...     ...  
278461  2019.0   10.0   6.0  20.0    45.0  
278462  2019.0   10.0   6.0  21.0    39.0  
278463  2019.0   10.0   6.0  22.0    18.0  
278464  2019.0   10.0   6.0  22.0    30.0  
278465  2019.0   10.0   6.0  22.0    30.0  

[278466 rows x 10 columns]

Drop InvoiceDate column after formation of Date, Year, Month, Day, Date, Hour & Minute from InvoiceDate columns

In [6]:
data.drop(["InvoiceDate"], axis=1, inplace=True)

Rearrange of columns as our choices

In [7]:
data = data[['Date','Year', 'Month','Day','Hour','Minute','ProductType','ItineraryType','NetFare']]
data.head()

Date    Year  Month   Day  Hour  Minute    ProductType ItineraryType  \
0  2018-01-12  2018.0    1.0  12.0  11.0    43.0  Other Product      Domestic   
1  2018-01-12  2018.0    1.0  12.0  11.0    43.0  Other Product      Domestic   
2  2018-01-12  2018.0    1.0  12.0  13.0     1.0            Air      Domestic   
3  2018-01-12  2018.0    1.0  12.0  13.0    10.0            Air      Domestic   
4  2018-11-30  2018.0   11.0  30.0  14.0    30.0            Air      Domestic   

   NetFare  
0  2160.00  
1     0.49  
2  6550.00  
3  6550.00  
4  5497.00

Converting data types on suitable format as mention below

In [8]:
#data['Year']=pd.to_datetime(data['float'])
data['Year']=data['Year'].astype('float')
data['Month']=data['Month'].astype('float')
data['Day']=data['Day'].astype('float')
data['Hour']=data['Hour'].astype('float')
data['Minute']=data['Minute'].astype('float')

Parse strings to datetime type 

In [9]:
data['Date'] = pd.to_datetime(data.Date, infer_datetime_format=True)
df=data.set_index(['Date'])

Data index sorted

In [10]:
df.sort_values(by='Date').head()

Year  Month  Day  Hour  Minute ProductType ItineraryType  \
Date                                                                     
2018-01-04  2018.0    1.0  4.0   8.0    26.0         Air      Domestic   
2018-01-04  2018.0    1.0  4.0   9.0    17.0         Air      Domestic   
2018-01-04  2018.0    1.0  4.0  22.0    38.0         Air      Domestic   
2018-01-04  2018.0    1.0  4.0  21.0    12.0         Air      Domestic   
2018-01-04  2018.0    1.0  4.0  20.0    59.0         Air      Domestic   

            NetFare  
Date                 
2018-01-04   8222.0  
2018-01-04   3418.0  
2018-01-04   6261.0  
2018-01-04   3473.0  
2018-01-04   5011.0

## Data Preparation <a id="4"></a>

<h4>Evaluating for Missing Data</h4>
There are two methods to detect missing data:
<ol>
 <li><b>.isnull()</b></li>
 <li><b>.notnull()</b></li>
</ol>   

Find missing value by Boolean method:
**"True"** stands for missing value, while **"False"** stands for not missing value.

In [11]:
missing_value = df.isnull()
missing_value

Year  Month    Day   Hour  Minute  ProductType  ItineraryType  \
Date                                                                         
2018-01-12  False  False  False  False   False        False          False   
2018-01-12  False  False  False  False   False        False          False   
2018-01-12  False  False  False  False   False        False          False   
2018-01-12  False  False  False  False   False        False          False   
2018-11-30  False  False  False  False   False        False          False   
...           ...    ...    ...    ...     ...          ...            ...   
2019-10-06  False  False  False  False   False        False          False   
2019-10-06  False  False  False  False   False        False          False   
2019-10-06  False  False  False  False   False        False          False   
2019-10-06  False  False  False  False   False        False          False   
2019-10-06  False  False  False  False   False        False          False   

            NetFare  
Date                 
2018-01-12    False  
2018-01-12    False  
2018-01-12    False  
2018-01-12    False  
2018-11-30    False  
...             ...  
2019-10-06    False  
2019-10-06    False  
2019-10-06    False  
2019-10-06    False  
2019-10-06    False  

[278466 rows x 8 columns]

### Count missing values in each column

Using a for loop in Python, we can quickly figure out the number of missing values in each column.<br>
**"True" stands for missing value, while "False" stands for not missing value**

In [12]:
for column in missing_value.columns.values.tolist():
    print(column)
    print (missing_value[column].value_counts())
    print("") 

Year
False    278464
True          2
Name: Year, dtype: int64

Month
False    278464
True          2
Name: Month, dtype: int64

Day
False    278464
True          2
Name: Day, dtype: int64

Hour
False    278464
True          2
Name: Hour, dtype: int64

Minute
False    278464
True          2
Name: Minute, dtype: int64

ProductType
False    278464
True          2
Name: ProductType, dtype: int64

ItineraryType
False    245689
True      32777
Name: ItineraryType, dtype: int64

NetFare
False    217576
True      60890
Name: NetFare, dtype: int64



Based on the summary above, each column has 278466 rows of data, all columns containing missing data:
<ol>
<li>"Year","Month","Day","Hour","Minute": 2 missing data</li>
<li>"ProductType": 2 missing data</li>
<li>"ItineraryType": 32777 missing data</li>    
<li>"NetFare": 60890 missing data</li>
</ol>

<h3 id="deal_missing_values">Deal with missing data</h3>

<b>Drop rows:</b>
<ol>
    <li>"Year","Month","Day","Hour","Minute": 2 missing data, drop them , two observations among large dataset are not so big deals</li>
    <li>"ProductType": 2 missing data, drop them , two observations among large dataset are not so big deals</li>
    <li>"ItineraryType": 32777 missing data, drop them , they are totally empty, we have huge data</li>
    <li>"NetFare": 60890 missing data, drop whole rows, they are totally empty</li>  
</ol>

Show data types

Drop missing values **["Year","Month","Day","Hour","Minute"]** in observations, the two rows are negelegible

In [13]:
df.dropna(subset=["Year","Month","Day","Hour","Minute"], axis=0, inplace=True)

Drop missing values **[ProductType]** in observations, the two rows are negelegible 

In [14]:
df.dropna(subset=["ProductType"], axis=0, inplace=True)

Drop missing values **[ItineraryType]** in observations, they are not very sensitive 

In [15]:
df.dropna(subset=["ItineraryType"], axis=0, inplace=True)

Drop missing values **[NetFare]** in observations, they are completly empty 

In [16]:
df.dropna(subset=["NetFare"], axis=0, inplace=True)

Reset index, because we droped many rows & sorted

In [17]:
df.reset_index().sort_values(by='Date').head()

Date    Year  Month  Day  Hour  Minute ProductType ItineraryType  \
93239 2018-01-04  2018.0    1.0  4.0  17.0    28.0         Air      Domestic   
93227 2018-01-04  2018.0    1.0  4.0  13.0    25.0         Air      Domestic   
93250 2018-01-04  2018.0    1.0  4.0  22.0    38.0         Air      Domestic   
93249 2018-01-04  2018.0    1.0  4.0  21.0    12.0         Air      Domestic   
93248 2018-01-04  2018.0    1.0  4.0  20.0    59.0         Air      Domestic   

       NetFare  
93239   7515.0  
93227   5168.0  
93250   6261.0  
93249   3473.0  
93248   5011.0

**Final Clean** data

In [18]:
df_clean = df.isnull()

**"True" stands for missing value, while "False" stands for not missing value**

In [19]:
for column in df_clean.columns.values.tolist():
    print(column)
    print (df_clean[column].value_counts())
    print("") 

Year
False    184800
Name: Year, dtype: int64

Month
False    184800
Name: Month, dtype: int64

Day
False    184800
Name: Day, dtype: int64

Hour
False    184800
Name: Hour, dtype: int64

Minute
False    184800
Name: Minute, dtype: int64

ProductType
False    184800
Name: ProductType, dtype: int64

ItineraryType
False    184800
Name: ItineraryType, dtype: int64

NetFare
False    184800
Name: NetFare, dtype: int64



RAW Data observations=**278466**, CLEANED Data observations=**184800**

#### Identifying Outliers with Interquartile Range (IQR)

In [20]:
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR = Q3 - Q1  
print(IQR)

Year          1.0
Month         6.0
Day          18.0
Hour          6.0
Minute       30.0
NetFare    4314.0
dtype: float64


The code below generates an output with the 'True' and 'False' values. Points where the values are 'True' represent the presence of the outlier.

In [21]:
print((df  < (Q1 - 1.5 * IQR)) | (df  > (Q3 + 1.5 * IQR)))

              Day   Hour  ItineraryType  Minute  Month  NetFare  ProductType  \
Date                                                                           
2018-01-12  False  False          False   False  False    False        False   
2018-01-12  False  False          False   False  False    False        False   
2018-01-12  False  False          False   False  False    False        False   
2018-01-12  False  False          False   False  False    False        False   
2018-11-30  False  False          False   False  False    False        False   
...           ...    ...            ...     ...    ...      ...          ...   
2019-10-06  False  False          False   False  False    False        False   
2019-10-06  False  False          False   False  False    False        False   
2019-10-06  False  False          False   False  False    False        False   
2019-10-06  False  False          False   False  False    False        False   
2019-10-06  False  False          False 

↑ Data is highly right skewed (11.13). Max value is 497678.It is due to outliers. So, there has necessary to delete outliers. 

We use the **IQR method** to delete outliers from records

In [22]:
df_out = df[~((df < (Q1 - 1.5 * IQR)) |(df > (Q3 + 1.5 * IQR))).any(axis=1)]
print(df_out.shape)

(167132, 8)


In [23]:
df_out.reset_index().sort_values(by='Date')

Date    Year  Month  Day  Hour  Minute       ProductType  \
84394 2018-01-04  2018.0    1.0  4.0  17.0    50.0               Air   
84395 2018-01-04  2018.0    1.0  4.0  17.0    50.0               Air   
84375 2018-01-04  2018.0    1.0  4.0   8.0    26.0               Air   
84376 2018-01-04  2018.0    1.0  4.0   9.0    17.0               Air   
84378 2018-01-04  2018.0    1.0  4.0  12.0     0.0               Air   
...          ...     ...    ...  ...   ...     ...               ...   
63505 2019-12-05  2019.0   12.0  5.0  14.0    52.0               Air   
63506 2019-12-05  2019.0   12.0  5.0  14.0    57.0  Air Cancellation   
63507 2019-12-05  2019.0   12.0  5.0  14.0    58.0     Other Product   
63497 2019-12-05  2019.0   12.0  5.0  14.0    51.0               Air   
63583 2019-12-05  2019.0   12.0  5.0  17.0    49.0     Other Product   

      ItineraryType  NetFare  
84394      Domestic   3577.0  
84395      Domestic   3577.0  
84375      Domestic   8222.0  
84376      Domestic   3418.0  
84378      Domestic   3625.0  
...             ...      ...  
63505      Domestic   2305.0  
63506      Domestic  -3600.0  
63507      Domestic    127.0  
63497      Domestic   3150.0  
63583      Domestic    269.0  

[167132 rows x 9 columns]

Among CLEAN Data observations=**184800**, after delation of outlier REMAINDER Data observations=**167132** 

In [24]:
import warnings
warnings.filterwarnings('ignore')

In [25]:
bins=np.linspace(min(df_out["NetFare"]),max(df_out["NetFare"]),6)

In [26]:
bins

array([-4315. ,  -863.8,  2587.4,  6038.6,  9489.8, 12941. ])

In [27]:
group_names=['Very Low','Low','Medium','High','Very High']

In [28]:
df_out['NetFare-bins']=pd.cut(df_out['NetFare'],bins,labels=group_names, include_lowest=True)

In [29]:
df_out[['NetFare','NetFare-bins']].head(55)

NetFare NetFare-bins
Date                             
2018-01-12   2160.00          Low
2018-01-12      0.49          Low
2018-01-12   6550.00         High
2018-01-12   6550.00         High
2018-11-30   5497.00       Medium
2018-01-12   4818.00       Medium
2018-11-30   1768.00          Low
2018-01-12   7300.00         High
2018-11-30   7050.00         High
2018-01-12   7300.00         High
2018-01-12   7650.00         High
2018-01-12  12437.00    Very High
2018-01-12   2400.00          Low
2018-01-12   2400.00          Low
2018-01-12    180.00          Low
2018-01-12    180.00          Low
2018-01-12    275.00          Low
2018-01-12    275.00          Low
2018-01-12   3833.00       Medium
2018-01-12   3649.00       Medium
2018-01-12   3649.00       Medium
2018-02-12     10.00          Low
2018-02-12    269.00          Low
2018-02-12    269.00          Low
2018-02-12    269.00          Low
2018-02-12    269.00          Low
2018-02-12   2218.00          Low
2018-02-12   2218.00          Low
2018-02-12   5743.00       Medium
2018-02-12     10.00          Low
2018-02-12   2599.00       Medium
2018-02-12     10.00          Low
2018-03-12   3650.00       Medium
2018-03-12   2413.00          Low
2018-03-12     10.00          Low
2018-03-12   2575.00          Low
2018-03-12   2575.00          Low
2018-03-12     10.00          Low
2018-03-12     10.00          Low
2018-03-12    269.00          Low
2018-04-12   2250.00          Low
2018-03-12   1670.00          Low
2018-03-12   7350.00         High
2018-03-12   7350.00         High
2018-03-12   4440.00       Medium
2018-03-12   5000.00       Medium
2018-03-12   3648.00       Medium
2018-03-12   4900.00       Medium
2018-03-12   3198.00       Medium
2018-03-12   4202.00       Medium
2018-03-12    269.00          Low
2018-03-12      0.00          Low
2018-03-12   6700.00         High
2018-03-12   2998.00       Medium
2018-03-12     10.00          Low

In [30]:
# using dictionary map
rank_map = {'Very Low':-2,'Low':-1,'Medium':0,'High':1,'Very High':2}
df_out['NetFareFinal'] = df_out['NetFare-bins'].map(rank_map)
df_out.head(50)

Year  Month   Day  Hour  Minute    ProductType ItineraryType  \
Date                                                                         
2018-01-12  2018.0    1.0  12.0  11.0    43.0  Other Product      Domestic   
2018-01-12  2018.0    1.0  12.0  11.0    43.0  Other Product      Domestic   
2018-01-12  2018.0    1.0  12.0  13.0     1.0            Air      Domestic   
2018-01-12  2018.0    1.0  12.0  13.0    10.0            Air      Domestic   
2018-11-30  2018.0   11.0  30.0  14.0    30.0            Air      Domestic   
2018-01-12  2018.0    1.0  12.0  14.0     9.0            Air      Domestic   
2018-11-30  2018.0   11.0  30.0  14.0    30.0            Air      Domestic   
2018-01-12  2018.0    1.0  12.0  13.0    10.0            Air      Domestic   
2018-11-30  2018.0   11.0  30.0  17.0    30.0            Air      Domestic   
2018-01-12  2018.0    1.0  12.0  14.0    25.0            Air      Domestic   
2018-01-12  2018.0    1.0  12.0  14.0    26.0            Air      Domestic   
2018-01-12  2018.0    1.0  12.0  13.0    30.0          Hotel      Domestic   
2018-01-12  2018.0    1.0  12.0  15.0    15.0            Air      Domestic   
2018-01-12  2018.0    1.0  12.0  15.0    15.0            Air      Domestic   
2018-01-12  2018.0    1.0  12.0  15.0    25.0  Other Product      Domestic   
2018-01-12  2018.0    1.0  12.0  15.0    25.0  Other Product      Domestic   
2018-01-12  2018.0    1.0  12.0  15.0    25.0  Other Product      Domestic   
2018-01-12  2018.0    1.0  12.0  15.0    25.0  Other Product      Domestic   
2018-01-12  2018.0    1.0  12.0  17.0    35.0          Hotel      Domestic   
2018-01-12  2018.0    1.0  12.0  21.0     6.0            Air      Domestic   
2018-01-12  2018.0    1.0  12.0  21.0     6.0            Air      Domestic   
2018-02-12  2018.0    2.0  12.0  11.0    18.0  Other Product      Domestic   
2018-02-12  2018.0    2.0  12.0  11.0    18.0  Other Product      Domestic   
2018-02-12  2018.0    2.0  12.0  11.0    18.0  Other Product      Domestic   
2018-02-12  2018.0    2.0  12.0  11.0    18.0  Other Product      Domestic   
2018-02-12  2018.0    2.0  12.0  11.0    18.0  Other Product      Domestic   
2018-02-12  2018.0    2.0  12.0  15.0    15.0          Hotel      Domestic   
2018-02-12  2018.0    2.0  12.0  15.0    15.0          Hotel      Domestic   
2018-02-12  2018.0    2.0  12.0  22.0    21.0            Air      Domestic   
2018-02-12  2018.0    2.0  12.0  22.0    25.0  Other Product      Domestic   
2018-02-12  2018.0    2.0  12.0  23.0    24.0            Air      Domestic   
2018-02-12  2018.0    2.0  12.0  23.0    25.0  Other Product      Domestic   
2018-03-12  2018.0    3.0  12.0   8.0    49.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0   8.0    49.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0   9.0     1.0  Other Product      Domestic   
2018-03-12  2018.0    3.0  12.0   9.0    22.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0   9.0    22.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0   8.0    45.0  Other Product      Domestic   
2018-03-12  2018.0    3.0  12.0  11.0     8.0  Other Product      Domestic   
2018-03-12  2018.0    3.0  12.0  11.0     8.0  Other Product      Domestic   
2018-04-12  2018.0    4.0  12.0  11.0     8.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0  11.0     1.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0  11.0    52.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0  11.0    52.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0  12.0    15.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0  11.0    37.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0  13.0    18.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0  12.0    20.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0  16.0    10.0            Air      Domestic   
2018-03-12  2018.0    3.0  12.0  16.

### Create dummie variables of ['ProductType', 'ItineraryType']

In [31]:
dummy_df=pd.get_dummies(df_out, columns=['ProductType', 'ItineraryType'], drop_first=False)

In [32]:
dummy_df.rename(columns = {'ProductType_Air':'ProTypeAir','ProductType_Air Cancellation':'ProTypeAirCanc','ProductType_Air Debit Note':'ProTypeAirDebtNote','ProductType_Air Loss':'ProTypeAirLoss','ProductType_Hotel':'ProTypeHot','ProductType_Hotel Cancellation':'ProTypeHotCanc','ProductType_Hotel Debit Note':'ProTypeHotDebNote','ProductType_Hotel Loss':'ProTypeHotLoss','ProductType_Other Product':'ProTypeOthPro','ProductType_Other Product Cancellation':'ProTypeOthProCanc','ProductType_Other Product Debit Note':'ProTypeOthProDebNote','ItineraryType_Domestic':'IteTypeDomes','ItineraryType_International':'IteTypeInter'}, inplace=True)


In [33]:
dummy_df.drop(['NetFare','NetFare-bins'],axis=1,inplace=True)

In [34]:
dummy_df.dtypes

Year                     float64
Month                    float64
Day                      float64
Hour                     float64
Minute                   float64
NetFareFinal            category
ProTypeAir                 uint8
ProTypeAirCanc             uint8
ProTypeAirDebtNote         uint8
ProTypeAirLoss             uint8
ProTypeHot                 uint8
ProTypeHotCanc             uint8
ProTypeHotDebNote          uint8
ProTypeHotLoss             uint8
ProTypeOthPro              uint8
ProTypeOthProCanc          uint8
ProTypeOthProDebNote       uint8
IteTypeDomes               uint8
IteTypeInter               uint8
dtype: object

In [35]:
# converting dtypes using astype 
dummy_df["Year"]= dummy_df["Year"].astype(int) 
dummy_df["Month"]= dummy_df["Month"].astype(int) 
dummy_df["Day"]= dummy_df["Day"].astype(int) 
dummy_df["Hour"]= dummy_df["Hour"].astype(int) 
dummy_df["Minute"]= dummy_df["Minute"].astype(int) 
dummy_df["NetFareFinal"]= dummy_df["NetFareFinal"].astype(int) 
dummy_df["ProTypeAir"]= dummy_df["ProTypeAir"].astype(int)
dummy_df["NetFareFinal"]= dummy_df["NetFareFinal"].astype(int) 
dummy_df["ProTypeAirCanc"]= dummy_df["ProTypeAirCanc"].astype(int) 
dummy_df["ProTypeAirDebtNote"]= dummy_df["ProTypeAirDebtNote"].astype(int) 
dummy_df["ProTypeAirLoss"]= dummy_df["ProTypeAirLoss"].astype(int) 
dummy_df["ProTypeHot"]= dummy_df["ProTypeHot"].astype(int) 
dummy_df["ProTypeHotCanc"]= dummy_df["ProTypeHotCanc"].astype(int) 
dummy_df["ProTypeHotDebNote"]= dummy_df["ProTypeHotDebNote"].astype(int) 
dummy_df["ProTypeHotLoss"]= dummy_df["ProTypeHotLoss"].astype(int) 
dummy_df["ProTypeOthPro"]= dummy_df["ProTypeOthPro"].astype(int)
dummy_df["ProTypeOthProCanc"]= dummy_df["ProTypeOthProCanc"].astype(int) 
dummy_df["ProTypeAirLoss"]= dummy_df["ProTypeAirLoss"].astype(int) 
dummy_df["ProTypeOthProDebNote"]= dummy_df["ProTypeOthProDebNote"].astype(int) 
dummy_df["IteTypeDomes"]= dummy_df["IteTypeDomes"].astype(int)
dummy_df["IteTypeInter"]= dummy_df["IteTypeInter"].astype(int)

In [36]:
data_model=dummy_df[['Month','Day','Hour','Minute','ProTypeAir','ProTypeHot','IteTypeDomes','NetFareFinal']]

In [37]:
data_model[100:150]

Month  Day  Hour  Minute  ProTypeAir  ProTypeHot  IteTypeDomes  \
Date                                                                         
2018-04-12      4   12    18       5           1           0             1   
2018-04-12      4   12    18       4           1           0             1   
2018-04-12      4   12    18       5           1           0             1   
2018-04-12      4   12    18       4           1           0             1   
2018-04-12      4   12    21      24           1           0             1   
2018-04-12      4   12    19      13           1           0             1   
2018-04-12      4   12    20      15           1           0             1   
2018-04-12      4   12    20      25           0           0             1   
2018-04-12      4   12    20      25           0           0             1   
2018-04-12      4   12    20      51           0           0             1   
2018-04-12      4   12    21      19           1           0             1   
2018-04-12      4   12    21      19           1           0             1   
2018-04-12      4   12    22      59           1           0             1   
2018-04-12      4   12    23       0           1           0             1   
2018-05-12      5   12     9      59           0           0             1   
2018-05-12      5   12    10       1           1           0             1   
2018-05-12      5   12    11      56           1           0             0   
2018-05-12      5   12    14      13           1           0             1   
2018-05-12      5   12    14      11           1           0             1   
2018-05-12      5   12    14      18           1           0             1   
2018-05-12      5   12    14      33           0           0             1   
2018-05-12      5   12    14      40           1           0             1   
2018-05-12      5   12    15      15           0           1             1   
2018-05-12      5   12    17      34           1           0             1   
2018-05-12      5   12    18       5           0           0             1   
2018-05-12      5   12    18       5           0           0             1   
2018-05-12      5   12    18       7           1           0             1   
2018-05-12      5   12    16      41           1           0             1   
2018-05-12      5   12    16      42           1           0             1   
2018-05-12      5   12    19       4           0           0             1   
2018-05-12      5   12    21      29           1           0             1   
2018-05-12      5   12    21      37           0           0             1   
2018-05-12      5   12    21      46           0           1             1   
2018-06-12      6   12    13       9           0           0             1   
2018-06-12      6   12    13       9           1           0             1   
2018-06-12      6   12    13       9           1           0             1   
2018-06-12      6   12    11      23           1           0             1   
2018-06-12      6   12    11      25           0           0             1   
2018-06-12      6   12    11      25           0           0             1   
2018-06-12      6   12    11      26           1           0             0   
2018-06-12      6   12    15      49           0           0             1   
2018-06-12      6   12    15      49           0           0             1   
2018-06-12      6   12    15      49           0           0             1   
2018-06-12      6   12    15      49           0           0             1   
2018-06-12      6   12    15      48           1           0             1   
2018-06-12      6   12    14      27           1           0             1   
2018-06-12      6   12    13      27           1           0             1   
2018-06-12      6   12    17      15           0           0             0   
2018-06-12      6   12    13      37           0           0             1   
2018-06-12      6   12    13      

### Setting up Environment<a id="5"></a>

In [38]:
import pycaret

In [39]:
from pycaret.regression import *
env= setup(data = data_model, target = 'NetFareFinal', session_id=2610,normalize = True, transformation = True, transform_target = True, 
                  combine_rare_levels = True, rare_level_threshold = 0.05,
                  remove_multicollinearity = True, multicollinearity_threshold = 0.95)

 
Setup Succesfully Completed!


### Create Model<a id="7"></a>

Create Linear Regression model

In [40]:
lr= create_model('lr',fold=10, round = 4)

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.5538  0.6124  0.7825  0.2208  0.4349  0.0398
1     0.5453  0.5963  0.7722  0.2236  0.4301  0.0219
2     0.5376  0.5864  0.7658  0.2324  0.4285  0.0545
3     0.5500  0.6029  0.7765  0.2344  0.4359  0.0531
4     0.5456  0.6014  0.7755  0.2236  0.4320  0.0465
5     0.5519  0.6103  0.7812  0.2364  0.4379  0.0709
6     0.5515  0.6111  0.7817  0.2268  0.4356  0.0432
7     0.5528  0.6108  0.7815  0.2276  0.4379  0.0633
8     0.5476  0.6079  0.7797  0.2226  0.4324  0.0295
9     0.5462  0.5956  0.7717  0.2375  0.4331  0.0677
Mean  0.5482  0.6035  0.7768  0.2286  0.4338  0.0490
SD    0.0046  0.0082  0.0053  0.0058  0.0030  0.0152

In [41]:
lr= finalize_model(lr)

In [42]:
print(lr)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)


In [43]:
predict_model(lr);

Model     MAE     MSE    RMSE      R2   RMSLE    MAPE
0  Linear Regression  0.5458  0.5979  0.7733  0.2277  0.4327  0.0474

## Saving the Model<a id="13"></a>

In [44]:
save_model(lr,'final_model_27Jun2020')

Transformation Pipeline and Model Succesfully Saved


"""
	5	12	10	1	1	0	1	0 <br>
	6	12	13	37	0	0	1	-1<br>
	5	12	14	13	1	0	1	2<br>
	5	12	14	11	1	0	1	1<br>
	5	12	14	18	1	0	1	0<br>
"""

<h4 style="color:red;">UI (User input) - Deployment</h4>

![](vl-vh_2.png)

In [45]:
import pickle

In [46]:
with open('final_model_27Jun2020.pkl', 'wb') as file:
    pickle.dump(lr, file)

In [47]:
with open('final_model_27Jun2020.pkl', 'rb') as file:
    ml_reg=pickle.load(file)

In [48]:
ml_reg.predict([[5,12,10,1,1,0,1]])

array([0.17811908])

In [49]:
import json,requests

In [50]:
data_model[1:4]

Month  Day  Hour  Minute  ProTypeAir  ProTypeHot  IteTypeDomes  \
Date                                                                         
2018-01-12      1   12    11      43           0           0             1   
2018-01-12      1   12    13       1           1           0             1   
2018-01-12      1   12    13      10           1           0             1   

            NetFareFinal  
Date                      
2018-01-12            -1  
2018-01-12             1  
2018-01-12             1

In [ ]:
url='http://localhost:9000/api'
data=json.dumps({'Month':1,'Day':12,'Hour':11,'Minute':43,'ProTypeAir':0,'ProTypeHot':0,'IteTypeDomes':1})
r=requests.post(url,data)
r.json()